src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7130,24 +7140,187 @@ public override IOperation VisitArgument(IArgumentOperation operation, int? capt
+
+                        var extraValueCaptureRef = new FlowCaptureReferenceOperation(extraValueCaptureId, operation.Operand.Syntax,
+                            operation.Operand.Type, constantValue: operation.Operand.GetConstantValue());
+
+                        assignment = makeAssignment(property, extraValueCaptureRef, operation);
+                        initializerBuilder.Add(assignment);
+                    }
+                }
+                LeaveRegionsUpTo(outerCaptureRegion);
+
+                return new AnonymousObjectCreationOperation(initializerBuilder.ToImmutableAndFree(), semanticModel: null, operation.Syntax, operation.Type, operation.IsImplicit);
+            }
 
-            IOperation cloned = operation.CloneMethod is null
-                ? MakeInvalidOperation(visitedInstance.Type, visitedInstance)
-                : new InvocationOperation(operation.CloneMethod, visitedInstance,
-                    isVirtual: true, arguments: ImmutableArray<IArgumentOperation>.Empty,
+            // Build an operation for `<implicitReceiver>.Property = <capturedValue>`
+            SimpleAssignmentOperation makeAssignment(IPropertySymbol property, IOperation capturedValue, WithOperation operation)
+            {
+                // <implicitReceiver>
+                var implicitReceiver = new InstanceReferenceOperation(InstanceReferenceKind.ImplicitReceiver,
                     semanticModel: null, operation.Syntax, operation.Type, isImplicit: true);
 
-            return PopStackFrame(frame, HandleObjectOrCollectionInitializer(operation.Initializer, cloned));
+                // <implicitReceiver>.Property
+                var target = new PropertyReferenceOperation(property, ImmutableArray<IArgumentOperation>.Empty, implicitReceiver,
+                    semanticModel: null, operation.Syntax, property.Type, isImplicit: true);
+
+                // <implicitReceiver>.Property = <capturedValue>
+                return new SimpleAssignmentOperation(isRef: false, target, capturedValue,
+                    semanticModel: null, operation.Syntax, property.Type, constantValue: null, isImplicit: true);
+            }
+
+            static bool setsAllProperties(ImmutableArray<IOperation> initializers, IEnumerable<IPropertySymbol> properties)
+            {
+                var set = new HashSet<IPropertySymbol>(SymbolEqualityComparer.IgnoreAll);
+                foreach (var property in properties)
+                {
+                    _ = set.Add(property);
+                }
+
+                foreach (var initializer in initializers)
+                {
+                    if (initializer is ISimpleAssignmentOperation simpleAssignment &&
+                        simpleAssignment.Target is not InvalidOperation)
+                    {
+                        var propertyReference = (IPropertyReferenceOperation)simpleAssignment.Target;
+                        _ = set.Remove(propertyReference.Property);
```

Comment:
>set.Remove(propertyReference.Property); [](http://example.com/codeflow?start=28&length=39)

It looks like we are making an assumption that a property that we get here belongs to the Anonymous Type that is cloned (otherwise we might be allocating captures that won't be used at the end). I think it is a valid assumption to make, but it might be worth adding an assert about that several lines above, where we asert other aspects of each property reference. Based on that assumption, consider if it is worth reducing the amount of HashSet operations (additions/removals) that we perform. For example, instead of first adding all properties into the set and then removing those that we run into, we could start with an empty set, add properties that we run into and at the end check if the count of items in the set equals to the count of properties in the type.